aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h

```diff
@@ -0,0 +1,522 @@
+    // If both are coalesced, search into the larger tensor is faster.
+    // Same holds when both are non-coalesced.
+    else {
+      Tensor larger, smaller;
+      std::tie(larger, smaller) = [&]() -> std::tuple<Tensor, Tensor> {
+        return x._nnz() >= y._nnz()
+          ? std::make_tuple(x, y)
+          : std::make_tuple(y, x);
+      }();
+
+      // If under a uniform distribution it is likely to hit many elements in larger,
+      // it is best to coalesce it for better performance.
+      const auto larger_sizes = larger.sizes();
+      const auto sparse_dim_numel = std::accumulate(
+          larger_sizes.begin(),
+          larger_sizes.begin() + larger.sparse_dim(),
+          1,
+          std::multiplies<int64_t>());
+      // If nnz > prod(larger.shape[:sparse_dim]), by the pidgeonhole principle,
+      // there is at least one bucket with nnz / prod(larger.shape[:sparse_dim]) elements.
+      // It provides a lower bound for the max count in the intersection.
+      // This condition is very conservative as we do not check whether such an event
+      // actually occurred, although it is very likely under a uniform distribution,
+      // the distribution with the highest uncertainty (maximizes entropy).
+      const auto max_count_lower_bound = larger._nnz() / sparse_dim_numel;
+      constexpr int64_t MAX_COPIES_PER_THREAD = 50;
+      return max_count_lower_bound > MAX_COPIES_PER_THREAD
+        ? std::make_tuple(larger.coalesce(), smaller)
+        : std::make_tuple(larger, smaller);
+    }
+  }();
+
+  // The employed hash function maps a d-dim index to a linear offset
+  // into a contiguous memory that is sufficient to fit a dense tensor
+  // of shape broadcasted_shape(x.shape, y.shape), i.e.
+  // idx -> \sum_{i = 0}^d idx[i] * hash_coeffs[i], where
+  // hash_coeffs are the strides of a contiguous tensor of shape
+  // broadcasted_shape(x.shape, y.shape).
+  // Assuming the following order on the dimensions, i.e. the right-most dim is the
+  // fastest-changing dim, and the left-most is the slowest-changing dim,
+  // which is implicit in the definition of hash_coeffs,
+  // it could be shown that the hash function is actually bijective and, hence,
+  // is a perfect hash function (no collisions ever).
+  const auto kHash = std::is_same<hash_t, int64_t>::value ? kLong : kInt;
+  const auto hash_coeffs = [&]() -> Tensor {
+    const auto broadcasted_sparse_dim_shape = std::vector<int64_t>(
+      broadcasted_shape.begin(),
+      broadcasted_shape.begin() + probably_coalesced.sparse_dim()
+    );
+    auto strides = contiguous_strides(broadcasted_sparse_dim_shape);
+    auto strides_len = static_cast<int64_t>(strides.size());
+    auto hash_coeffs = at::empty(
+        {strides_len},
+        probably_coalesced._indices().options().device(kCPU).dtype(kHash));
+    // Copy with a potential casting. Is there a nicer way?
+    for (const auto i : c10::irange(strides_len)) {
```

Comment:
This likely has a lot of overhead, since you're repeated indexing into a small Tensor on CPU. Based on the use site `ptr_hash_coeffs[dim]` I'm surprised this needs to be a Tensor at all (and not say a SmallVector) or even live on the device of `probably_coalesced` (which incurs an expensive host to device transfer).

Reply:
Does SmallVector work on CUDA? What would you suggest as a better alternative for moving stride values onto a device?

Reply:
Ok, I missed the exact context in which this Tensor is being used... Indeed you'll need to construct a Tensor to then use that data on the GPU.